zkevm-circuits/src/bytecode_circuit/circuit.rs

```diff
@@ -0,0 +1,1030 @@
+            );
+
+            cb.require_equal(
+                "next.hash == cur.hash",
+                meta.query_advice(bytecode_table.code_hash, Rotation::next()),
+                meta.query_advice(bytecode_table.code_hash, Rotation::cur()),
+            );
+
+            // TODO: check this
+            cb.require_equal(
+                "next.value_rlc == cur.value_rlc * randomness + next.value",
+                meta.query_advice(value_rlc, Rotation::next()),
+                meta.query_advice(value_rlc, Rotation::cur()) * challenges.keccak_input()
+                    + meta.query_advice(value, Rotation::next()),
+            );
+
+            cb.require_equal(
+                "next.push_data_left == cur.is_code ? cur.push_data_size : cur.push_data_left - 1",
+                meta.query_advice(push_data_left, Rotation::next()),
+                select::expr(
+                    meta.query_advice(bytecode_table.is_code, Rotation::cur()),
+                    meta.query_advice(push_data_size, Rotation::cur()),
+                    meta.query_advice(push_data_left, Rotation::cur()) - 1.expr(),
+                ),
+            );
+
+            cb.gate(and::expr(vec![
+                meta.query_fixed(q_enable, Rotation::cur()),
+                not::expr(meta.query_fixed(q_last, Rotation::cur())),
+                is_byte_to_byte(meta),
+            ]))
+        });
+
+        // When is_byte_to_header ->
+        // assert cur.index + 1 == cur.length
+        // assert keccak256_table_lookup(cur.hash, cur.length, cur.value_rlc)
+        meta.create_gate("Byte to Header row", |meta| {
+            let mut cb = BaseConstraintBuilder::default();
+
+            cb.require_equal(
+                "cur.index + 1 == cur.length",
+                meta.query_advice(bytecode_table.index, Rotation::cur()) + 1.expr(),
+                meta.query_advice(length, Rotation::cur()),
+            );
+
+            cb.gate(and::expr(vec![
+                meta.query_fixed(q_enable, Rotation::cur()),
+                not::expr(meta.query_fixed(q_last, Rotation::cur())),
+                is_byte_to_header(meta),
+            ]))
+        });
+        meta.lookup_any(
+            "keccak256_table_lookup(cur.hash, cur.length, cur.value_rlc)",
+            |meta| {
+                let enable = and::expr(vec![
+                    meta.query_fixed(q_enable, Rotation::cur()),
+                    not::expr(meta.query_fixed(q_last, Rotation::cur())),
+                    is_byte_to_header(meta),
+                ]);
+
+                let lookup_columns = vec![value_rlc, length, bytecode_table.code_hash];
+
+                let mut constraints = vec![(
+                    enable.clone(),
+                    meta.query_advice(keccak_table.is_enabled, Rotation::cur()),
+                )];
+
+                // TODO: perhaps write this explicitly so it is more readable the matching
```

Comment:
Agree we should make this more readable. One idea I have in mind is to have a function in keccak table side.
`match_columns(input_rlc, length, output_hash) -> Vec<(Column, Column)>`
The function is only responsible for yielding correct pairs of columns. We then add selectors and query cells here.

Reply:
I was going to complain, but I actually think what you are saying makes sense.